README.md
# MonoHook

Expose a single HTTP webhook endpoint that executes a command:

```
$> monohook --authorization letmein --concurrency 1 --port 5000 -- deploy.sh
```

Trigger execution of the command by POSTing to the webhook:

```
$> curl -v -X POST -H "Authorization: Bearer letmein" http://localhost:5000
> POST / HTTP/1.1
> Host: localhost:5000
> User-Agent: curl/7.54.0
> Accept: */*
> Authorization: Bearer letmein
>
< HTTP/1.1 202 Accepted
< Date: Wed, 15 May 2019 07:06:30 GMT
< Content-Length: 0
```

The command is executed asynchronously.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Installation](#installation)
  - [Docker](#docker)
  - [Download binary](#download-binary)
- [Usage](#usage)
  - [Authentication](#authentication)
  - [Concurrency](#concurrency)
  - [Buffering](#buffering)
  - [Command environment & working directory](#command-environment--working-directory)
  - [Forwarding HTTP request data to the command](#forwarding-http-request-data-to-the-command)
  - [Port number](#port-number)
  - [TL;DR](#tldr)
- [Exit codes](#exit-codes)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

* [Changelog](CHANGELOG.md)

[![version](https://img.shields.io/badge/Version-v3.0.1-blue.svg)](https://github.com/AlphaHydrae/monohook/releases/tag/v3.0.1)
[![build status](https://travis-ci.org/AlphaHydrae/monohook.svg?branch=master)](https://travis-ci.org/AlphaHydrae/monohook)
[![license](https://img.shields.io/badge/License-MIT-blue.svg)](LICENSE.txt)



## Installation

### Docker

MonoHook is [available on Docker Hub][hub]. Run it with Docker:

```
$> docker run --rm -it -p 5000:5000 alphahydrae/monohook:2.0.0 -a letmein -- echo Hook triggered
```

Test it:

```
$> curl -v -X POST -H "Authorization: Bearer letmein" http://localhost:5000
```

You should see the command being executed in the container's logs:

```
Executing /bin/echo
Hook triggered
Successfully executed /bin/echo
```

To use MonoHook in a `Dockerfile`, simply download the binary and make it executable:

```
RUN wget -O /usr/local/bin/monohook \
    https://github.com/AlphaHydrae/monohook/releases/download/v3.0.1/monohook_linux_amd64 && \
    chmod +x /usr/local/bin/monohook
```

### Download binary

* **Linux**

  ```
  wget -O /usr/local/bin/monohook \
    https://github.com/AlphaHydrae/monohook/releases/download/v3.0.1/monohook_linux_amd64 && \
    chmod +x /usr/local/bin/monohook
  ```
* **Linux (arm64)**

  ```
  wget -O /usr/local/bin/monohook \
    https://github.com/AlphaHydrae/monohook/releases/download/v3.0.1/monohook_linux_arm64 && \
    chmod +x /usr/local/bin/monohook
  ```
* **macOS**

  ```
  wget -O /usr/local/bin/monohook \
    https://github.com/AlphaHydrae/monohook/releases/download/v3.0.1/monohook_darwin_amd64 && \
    chmod +x /usr/local/bin/monohook
  ```
* **Windows**

  ```
  wget -O /usr/local/bin/monohook \
    https://github.com/AlphaHydrae/monohook/releases/download/v3.0.1/monohook_windows_amd64 && \
    chmod +x /usr/local/bin/monohook
  ```



## Usage

MonoHook listens on a given TCP port, 5000 by default, and executes a command
when a request is received. The command must be specified after the argument
terminator (`--`):

```
$> monohook -- echo Hook triggered
```

To trigger execution of the command, simply make a POST request to the hook:

```
$> curl -v -X POST http://localhost:5000
```

If you go back to the terminal where MonoHook is running, you will see that the
command's output is piped to MonoHook's own standard output so you can see it:

```
Executing /bin/echo
Hook triggered
Successfully executed /bin/echo
```

### Authentication

You can secure the webhook by requiring that an authentication token be sent,
using the `-a, --authorization <TOKEN>` command-line flag or the
`$MONOHOOK_AUTHORIZATION` environment variable:

```
$> monohook -a letmein -- echo Hook triggered
```

MonoHook will respond to requests without authentication with HTTP 401
Unauthorized, and the command will not be executed:

```
$> curl -v -X POST http://localhost:5000
> POST / HTTP/1.1
> Host: localhost:5000
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 401 Unauthorized
< Date: Wed, 15 May 2019 07:22:05 GMT
< Content-Length: 0
```

You may send the authentication token as a bearer token in the `Authorization`
header:

```
$> curl -v -X POST -H "Authorization: Bearer letmein" http://localhost:5000
> POST / HTTP/1.1
> Host: localhost:5000
> User-Agent: curl/7.54.0
> Accept: */*
> Authorization: Bearer letmein
>
< HTTP/1.1 202 Accepted
< Date: Wed, 15 May 2019 07:24:11 GMT
< Content-Length: 0
```

Or in the `authorization` URL query parameter:

```
$> curl -v -X POST "http://localhost:5000?authorization=letmein"
> POST /?authorization=letmein HTTP/1.1
> Host: localhost:5000
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 202 Accepted
< Date: Wed, 15 May 2019 07:24:59 GMT
< Content-Length: 0
```

> MonoHook does not support HTTPS, therefore it is recommended that you put it
> behind a reverse proxy like Apache or nginx and use an SSL/TLS certificate.
>
> For example, you could use this location in an nginx server block:
>
> ```
> location /hooks/deploy {
>   proxy_set_header X-Real-IP $remote_addr;
>   proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
>   proxy_set_header X-Forwarded-Proto https;
>   proxy_set_header Host $http_host;
>   proxy_redirect off;
>   proxy_pass http://localhost:3000;
> }
> ```

### Concurrency

By default, MonoHook will allow only one execution of the command at a time. If
multiple requests are received in quick succession, they will be buffered until
each previou execution has completed.

You may test this by running MonoHook with no options:

```
$> monohook -- echo Hook triggered
```

Then triggering 10 requests with cURL:

```
$> for n in 0 1 2 3 4 5 6 7 8 9; do curl -v -X POST http://localhost:5000; done
< HTTP/1.1 202 Accepted
< HTTP/1.1 202 Accepted
< HTTP/1.1 202 Accepted
...
```

You can see in MonoHook's output that it is waiting for each command execution
to complete before starting the next one:

```
Executing /bin/echo
Hook triggered
Successfully executed /bin/echo

Executing /bin/echo
Hook triggered
Successfully executed /bin/echo

Executing /bin/echo
Hook triggered
Successfully executed /bin/echo

...
```

You may specify a greater concurrency with the `-c, --concurrency <LIMIT>`
command-line flag, or the `$MONOHOOK_CONCURRENCY` environment variable. This
example uses the `sleep` command to simulate a long-running command:

```
$> monohook -c 3 -- sleep 2
```

Trigger 10 requests like before:

```
$> for n in 0 1 2 3 4 5 6 7 8 9; do curl -v -X POST http://localhost:5000; done
< HTTP/1.1 202 Accepted
< HTTP/1.1 202 Accepted
< HTTP/1.1 202 Accepted
...
```

This time, you will see MonoHook executing the command in parallel in batches of
3:

```
Executing /bin/sleep
Executing /bin/sleep
Executing /bin/sleep
Successfully executed /bin/sleep
Successfully executed /bin/sleep
Successfully executed /bin/sleep

Executing /bin/sleep
Executing /bin/sleep
Executing /bin/sleep
Successfully executed /bin/sleep
Successfully executed /bin/sleep
Successfully executed /bin/sleep

...
```

> You may specify infinite concurrency by setting a concurrency of zero:
> `monohook -c 0 -- sleep 2`.

### Buffering

If concurrency is limited (which is the default) and MonoHook is already busy
executing a long-running command or commands, it will keep a buffer of pending
requests to execute. They will be executed as soon as the execution of previous
commands has completed and the configured concurrency limit permits it.

This buffer has a limited size (10 by default). If the buffer is full and
MonoHook receives more requests, it will discard them and respond with HTTP 429
Too Many Requests. Those requests will never be processed.

You may observe this behavior by running MonoHook with a small buffer and a
long-running command:

```
$> monohook -b 3 -- sleep 2
```

Trigger 10 requests like before:

```
$> for n in 0 1 2 3 4 5 6 7 8 9; do curl -v -X POST http://localhost:5000; done
< HTTP/1.1 202 Accepted
< HTTP/1.1 202 Accepted
< HTTP/1.1 202 Accepted
< HTTP/1.1 202 Accepted
< HTTP/1.1 429 Too Many Requests
< HTTP/1.1 429 Too Many Requests
< HTTP/1.1 429 Too Many Requests
...
```

In this example, MonoHook immediately started executing the first request, then
put the next 3 in the buffer. Subsequent requests are discarded.

The size of the buffer can be configured with the `-b, --buffer <SIZE>`
command-line flag or the `$MONOHOOK_BUFFER` environment variable.

> You may specify an infinite buffer by setting the buffer to zero: `monohook -b
> 0 -- sleep 2`

### Command environment & working directory

The command will be executed with the same environment variables as those
available to MonoHook. By default, the working directory is the one where
MonoHook is run.

It can be configured to execute the command in a specific working directory with
the `-C, --cwd <DIR>` command-line flag or the `$MONOHOOK_CWD` environment
variable:

```
$> monohook -C /path/to/somewhere - pwd
```

### Forwarding HTTP request data to the command

The command does not have access to the HTTP request that triggered it by
default. You may choose to forward part or all of the following data:

* The HTTP request body with the `-B, --forward-request-body` command-line flag
  or the `$MONOHOOK_FORWARD_REQUEST_BODY` environment variable.

  If enabled, the request body will be piped to the command's standard input.

  If the `-Y, --cache-request-body` command-line flag or the
  `$MONOHOOK_CACHE_REQUEST_BODY` environment variable is also enabled, the
  request body will be read immediately and cached for when the hook executes
  (which might take a while depending on concurrency settings). Otherwise, the
  request will wait to complete until the hook can actually execute.
* HTTP request headers with the `-H, --forward-request-headers` command-line
  flag or the `$MONOHOOK_FORWARD_REQUEST_HEADERS` environment variable.

  If enabled, request headers will be provided to the command as environment
  variables, with the following naming convention:

      Accept -> $MONOHOOK_REQUEST_HEADER_ACCEPT
      Content-Type -> $MONOHOOK_REQUEST_HEADER_CONTENT_TYPE
* The HTTP request URL with the `-U, --forward-request-url` command-line flag or
  the `$MONOHOOK_FORWARD_REQUEST_URL` environment variable.

  If enabled, the URL parsed from the HTTP request line is provided to the
  command as the `$MONOHOOK_REQUEST_URL` environment variable.

The following example shows how to save the body of each HTTP request to a file:

```
$> monohook -B -- tee -a body.txt
```

Trigger 10 cURL requests with a body:

```
$> for n in 0 1 2 3 4 5 6 7 8 9; do curl -v -d $n http://localhost:5000; done
```

Check the file's contents:

```
$> cat body.txt
0123456789
```

> Note that forwarding the request body without caching it (using the `-Y,
> --cache-request-body` command-line flag or the `$MONOHOOK_CACHE_REQUEST_BODY`
> environment variable) means that the execution of the command will not end
> before the HTTP request body has been completely consumed. This may have a
> performance impact. For example, with a concurrency of 1, each request will
> have to wait for the previous one to upload its data.

### Port number

MonoHook listens on port 5000 by default. You may use a different port with the
`-p, --port <NUMBER>` command-line flag or the `$MONOHOOK_PORT` environment
variable.

### TL;DR

```
monohook exposes a single HTTP webhook endpoint that executes a command.

Usage:
  monohook [OPTION...] [--] [EXEC...]

Options:
  -a, --authorization string      Authentication token that must be sent as a Bearer token in the 'Authorization' header or as the 'authorization' URL query parameter
  -b, --buffer uint               Maximum number of requests to queue before refusing subsequent ones until the queue is freed (zero for infinite) (default 10)
  -Y, --cache-request-body        Whether to cache the HTTP request's body so the hook can return right away
  -c, --concurrency uint          Maximum number of times the command should be executed in parallel (zero for infinite concurrency) (default 1)
  -C, --cwd string                Working directory in which to run the command
  -B, --forward-request-body      Whether to forward each HTTP request's body to the the command's standard input
  -H, --forward-request-headers   Whether to forward each HTTP request's headers to the the command as environment variables (e.g. Content-Type becomes $MONOHOOK_REQUEST_HEADER_CONTENT_TYPE)
  -U, --forward-request-url       Whether to forward each HTTP request's URL to the the command as the $MONOHOOK_REQUEST_URL environment variable
  -p, --port uint                 Port on which to listen to (default 5000)
  -q, --quiet                     Do not print anything except the command's standard output and error streams (default false)

Examples:
  Update a file when the hook is triggered:
    monohook -- touch hooked.txt
  Deploy an application when the hook is triggered:
    monohook -a letmein -- deploy-stuff.sh
```



## Exit codes

**MonoHook** may exit with the following status codes to indicate a problem:

Code | Description
:--- | :---
`1`  | Invalid arguments were given.
`2`  | The command to execute (provided after `--`) could not be found in the `$PATH`.



[hub]: https://hub.docker.com/r/alphahydrae/monohook


CHANGELOG.md
# Changelog

## v3.1.0

* Support reading the request body immediately and caching it for when the hook
  actually executes, using the `-Y, --cache-request-body` command-line flag or
  the corresponding `$MONOHOOK_CACHE_REQUEST_BODY` environment variable.

## v3.0.1

* Fix the `-p, --port` command-line flag which had no effect.

## v3.0.0

* **Breaking:** environment variables are no longer interpolated in command-line
  flag values.
* Environment variables can be used instead of command-line flags (e.g.
  `$MONOHOOK_CONCURRENCY` for `--concurrency`). Command-line flags take
  precedence over environment variables.
* Allow forwarding the HTTP request body, headers and URL to the command to
  execute.

## v2.0.0

* **Breaking:** require `POST` requests instead of accepting any method.
* Allow authorization to be provided with the `authorization` URL query
  parameter in addition to the `Authorization` header.
* Add a Dockerfile and [Docker Hub automated
  builds](https://hub.docker.com/r/alphahydrae/monohook).

## v1.0.0

* Initial release.


utils/cli.go
package utils

import (
	"fmt"
	"os"
	"regexp"
	"strconv"

	"github.com/fatih/color"
	flag "github.com/spf13/pflag"
)

var falseRegexp = regexp.MustCompile("^(?:0|n|no|f|false)$")
var trueRegexp = regexp.MustCompile("^(?:1|y|yes|t|true)$")

// EnvPrefix is the prefix common to environment variables for monohook.
const EnvPrefix = "MONOHOOK_"

// ErrorHandler can be called if an error occurs while parsing a command line
// option.
type ErrorHandler func(code int, message string)

// Fail prints a message in red to the standard error stream (as long as the
// `quiet` option is false) and exits the process with a non-zero code.
func Fail(code int, quiet bool, format string, values ...interface{}) {
	if !quiet {
		Print(quiet, fmt.Sprintf(color.RedString("Error: "+format+"\n"), values...))
	}

	os.Exit(code)
}

// Print prints a message to the standard error stream (as long as the `quiet`
// option is false).
func Print(quiet bool, format string, values ...interface{}) {
	if !quiet {
		fmt.Fprintf(os.Stderr, format, values...)
	}
}

// BoolOption parses and returns the value of a boolean option, either from
// command line flags or from an environment variable.
func BoolOption(value *bool, longOpt string, shortOpt string, envVarName string, defaultVal bool, desc string, fail ErrorHandler) {

	defaultFlagVal := defaultVal

	envVar := EnvPrefix + envVarName
	envValue := os.Getenv(envVar)
	if trueRegexp.MatchString(envValue) {
		defaultFlagVal = true
	} else if falseRegexp.MatchString(envValue) {
		defaultFlagVal = false
	} else if envValue != "" {
		fail(1, fmt.Sprintf("environment variable $%s must be one of: 0, n, no, f, false, 1, y, yes, t, true", envVar))
	}

	flag.BoolVarP(value, longOpt, shortOpt, defaultFlagVal, desc)
}

// StringOption parses and returns the value of a free-form text option, either
// from command line flags or from an environment variable.
func StringOption(value *string, longOpt string, shortOpt string, envVarName string, defaultVal string, desc string) {

	defaultFlagVal := defaultVal

	envVar := EnvPrefix + envVarName
	envValue := os.Getenv(envVar)
	if envValue != "" {
		defaultFlagVal = envValue
	}

	flag.StringVarP(value, longOpt, shortOpt, defaultFlagVal, desc)
}

// Uint64Option parses and returns the value of an unsigned integer option,
// either from command line flags or from an environment variable.
func Uint64Option(value *uint64, longOpt string, shortOpt string, envVarName string, defaultVal uint64, desc string, fail ErrorHandler) {

	defaultFlagVal := defaultVal

	envVar := EnvPrefix + envVarName
	envValue := os.Getenv(envVar)
	if envValue != "" {

		parsed, err := strconv.ParseUint(envValue, 10, 64)
		if err != nil {
			fail(1, fmt.Sprintf("environment variable $%s must be an integer greater than or equal to zero", envVar))
		}

		defaultFlagVal = parsed
	}

	flag.Uint64VarP(value, longOpt, shortOpt, defaultFlagVal, desc)
}


scripts/build.sh
#!/usr/bin/env bash
set -e

bold=$(tput bold)
normal=$(tput sgr0)
versions="darwin_amd64 linux_amd64 linux_arm64 windows_amd64"

test -z "$RELEASE" && RELEASE="$(git describe --tags || echo -n)"

if test -z "$RELEASE"; then
  >&2 echo No Git tag found
  exit 1
fi

rm -fr release/${RELEASE}

printf "\n${bold}○ Building binaries...${normal}\n"

for version in $versions; do
  os="$(echo $version | cut -d _ -f 1)"
  arch="$(echo $version | cut -d _ -f 2)"
  env GOOS=$os GOARCH=$arch go build -ldflags="-s -w" -o release/${RELEASE}/monohook_${os}_${arch} &
done

wait

for version in $versions; do
  file="release/${RELEASE}/monohook_${version}"
  printf "\n${bold}○ Compressing ${file}...${normal}\n\n"
  upx --ultra-brute "$file"
done

printf "\n${bold}○ Calculating digests...${normal}\n\n"
dgstore 'release/**/*'

echo


monohook.go
// The monohook command runs a single HTTP webhook endpoint that executes a command.
package main

import (
	"bytes"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"sync"

	"github.com/alphahydrae/monohook/utils"
	"github.com/fatih/color"
	flag "github.com/spf13/pflag"
)

const usageHeader = `%s exposes a single HTTP webhook endpoint that executes a command.

Usage:
  %s [OPTION...] [--] [EXEC...]

Options:
`

const usageFooter = `
Examples:
  Update a file when the hook is triggered:
    monohook -- touch hooked.txt
  Deploy an application when the hook is triggered:
    monohook -a letmein -- deploy-stuff.sh
`

type commandOptions struct {
	command string
	args    []string
	cwd     string
	env     []string
	quiet   bool
	reader  *io.PipeReader
	cache   []byte
}

var httpRequestHeaderToEnvVarRegexp = regexp.MustCompile("-")

func main() {

	var auth string
	var buffer uint64
	var concurrency uint64
	var cwd string
	var forwardRequestBody bool
	var cacheRequestBody bool
	var forwardRequestHeaders bool
	var forwardRequestURL bool
	var port uint64

	quiet := false
	errHandler := func(code int, message string) {
		utils.Fail(code, quiet, message)
	}

	flag.CommandLine.SetOutput(os.Stdout)
	utils.BoolOption(&quiet, "quiet", "q", "QUIET", false, "Do not print anything except the command's standard output and error streams (default false)", errHandler)

	utils.StringOption(&auth, "authorization", "a", "AUTHORIZATION", "", "Authentication token that must be sent as a Bearer token in the 'Authorization' header or as the 'authorization' URL query parameter")
	utils.Uint64Option(&buffer, "buffer", "b", "BUFFER", 10, "Maximum number of requests to queue before refusing subsequent ones until the queue is freed (zero for infinite)", errHandler)
	utils.Uint64Option(&concurrency, "concurrency", "c", "CONCURRENCY", 1, "Maximum number of times the command should be executed in parallel (zero for infinite concurrency)", errHandler)
	utils.BoolOption(&forwardRequestBody, "forward-request-body", "B", "FORWARD_REQUEST_BODY", false, "Whether to forward each HTTP request's body to the the command's standard input", errHandler)
	utils.BoolOption(&cacheRequestBody, "cache-request-body", "Y", "CACHE_REQUEST_BODY", false, "Whether to cache the HTTP request's body so the hook can return right away", errHandler)
	utils.BoolOption(&forwardRequestHeaders, "forward-request-headers", "H", "FORWARD_REQUEST_HEADERS", false, "Whether to forward each HTTP request's headers to the the command as environment variables (e.g. Content-Type becomes $MONOHOOK_REQUEST_HEADER_CONTENT_TYPE)", errHandler)
	utils.BoolOption(&forwardRequestURL, "forward-request-url", "U", "FORWARD_REQUEST_URL", false, "Whether to forward each HTTP request's URL to the the command as the $MONOHOOK_REQUEST_URL environment variable", errHandler)
	utils.StringOption(&cwd, "cwd", "C", "CWD", "", "Working directory in which to run the command")
	utils.Uint64Option(&port, "port", "p", "PORT", 5000, "Port on which to listen to", errHandler)

	flag.Usage = func() {
		fmt.Printf(usageHeader, os.Args[0], os.Args[0])
		flag.PrintDefaults()
		fmt.Print(usageFooter)
	}

	flag.Parse()

	if cacheRequestBody && !forwardRequestBody {
		utils.Fail(1, quiet, "option -Y, --cache-request-body can only be used together with option -B, --forward-request-body")
	} else if port > 65535 {
		utils.Fail(1, quiet, "option -p, --port or environment variable $PORT must be an integer smaller than or equal to 65535")
	}

	terminator := -1
	for i := 0; i < len(os.Args); i++ {
		if os.Args[i] == "--" {
			terminator = i
			break
		}
	}

	var extra []string
	var execCommand string
	var execArgs []string
	if terminator >= 0 && terminator < len(os.Args)-1 {
		extra = flag.Args()[0 : len(flag.Args())-(len(os.Args)-terminator-1)]

		var err error
		execCommand, err = exec.LookPath(os.Args[terminator+1])
		if err != nil {
			utils.Fail(2, quiet, "could not find command \"%s\"", os.Args[terminator+1])
		}

		execArgs = os.Args[terminator+2 : len(os.Args)]
	} else {
		extra = flag.Args()
	}

	if len(extra) != 0 {
		utils.Fail(1, quiet, "no argument expected before the terminator (put -- before the command to execute)")
	} else if execCommand == "" {
		utils.Fail(1, quiet, "no command to execute was provided")
	}

	execCh := make(chan *commandOptions, buffer)

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {

		// Refuse non-POST requests
		if strings.ToUpper(r.Method) != "POST" {
			w.WriteHeader(405)
			return
		}

		// Refuse request if unauthorized.
		if !utils.Authorized(auth, r) {
			w.WriteHeader(401)
			return
		}

		opts := &commandOptions{}
		opts.command = execCommand
		opts.args = execArgs
		opts.cwd = cwd
		opts.quiet = quiet

		// Forward HTTP request body to the command's standard input.
		var pw *io.PipeWriter
		if forwardRequestBody {
			if cacheRequestBody {
				body, err := ioutil.ReadAll(r.Body)
				if err != nil {
					utils.Print(quiet, "Warning: could not read request body: %s\n", err)
					w.WriteHeader(400)
					return
				}

				opts.cache = body
			} else {
				reader, writer := io.Pipe()
				opts.reader = reader
				pw = writer
			}
		}

		var env []string
		env = append(env, os.Environ()...)

		// Forward HTTP request headers to the command as environment variables.
		if forwardRequestHeaders {
			for headerName, headerVals := range r.Header {
				env = append(env, utils.EnvPrefix+"REQUEST_HEADER_"+strings.ToUpper(httpRequestHeaderToEnvVarRegexp.ReplaceAllString(headerName, "_"))+"="+strings.Join(headerVals, ","))
			}
		}

		// Forward the HTTP request URL to the command as an environment variable
		if forwardRequestURL {
			env = append(env, utils.EnvPrefix+"REQUEST_URL="+r.URL.String())
		}

		opts.env = env

		// Refuse extra requests if buffer is full.
		select {
		case execCh <- opts:
			w.WriteHeader(202)
		default:
			w.WriteHeader(429)
		}

		if pw != nil {

			_, err := io.Copy(pw, r.Body)
			if err != nil {
				utils.Print(quiet, "Warning: could not read entire request body: %s\n", err)
			}

			pwErr := pw.Close()
			if pwErr != nil {
				utils.Print(quiet, "Warning: could not close request body pipe: %s\n", pwErr)
			}
		}
	})

	s := &http.Server{
		Addr: ":" + strconv.FormatUint(port, 10),
	}

	go worker(concurrency, quiet, execCh)

	s.ListenAndServe()
}

func worker(concurrency uint64, quiet bool, execChannel chan *commandOptions) {
	utils.Print(quiet, "Execution worker started\n")

	n := uint64(0)
	wait := &sync.WaitGroup{}

	for opts := range execChannel {
		if concurrency == 0 {
			go execCommand(opts, nil)
		} else {

			n++
			wait.Add(1)
			go execCommand(opts, wait)

			// Wait for queue to clear if concurrency is limited.
			if n >= concurrency {
				wait.Wait()
				n -= concurrency
				wait = &sync.WaitGroup{}
			}
		}
	}
}

func execCommand(opts *commandOptions, waitGroup *sync.WaitGroup) {

	utils.Print(opts.quiet, color.YellowString("Executing %s\n"), opts.command)

	cmd := exec.Command(opts.command, opts.args...)
	cmd.Dir = opts.cwd
	cmd.Env = opts.env
	cmd.Stderr = os.Stdout
	cmd.Stdout = os.Stdout

	if opts.cache != nil {
		cmd.Stdin = bytes.NewBuffer(opts.cache)
	} else if opts.reader != nil {
		cmd.Stdin = opts.reader
	}

	err := cmd.Run()
	if err != nil {
		utils.Print(opts.quiet, color.RedString("Command %s error: %s\n"), opts.command, err)
	} else {
		utils.Print(opts.quiet, color.GreenString("Successfully executed %s\n"), opts.command)
	}

	if opts.reader != nil {
		opts.reader.Close()
	}

	if waitGroup != nil {
		waitGroup.Done()
	}
}
